File: backend/src/Application/Services/System/UserService.cs
using Application.Contracts;
using Application.Services;
using Domain.Entities.Auth;
using System.ComponentModel.DataAnnotations;

namespace Application.Services.System
{
    public class UserService(IUnitOfWork unitOfWork, ILocalizationService localizationService, ILanguageCatalog languageCatalog) : IUserService
    {
        public async Task<IEnumerable<User>> GetAllUsers()
        {
            var users = await unitOfWork.Users.GetAll();
            return users.OrderBy(u => u.Username);
        }

        public async Task<User?> GetUserById(Guid id)
        {
            return await unitOfWork.Users.Get(id);
        }

        public async Task<GenericResponse> CreateUser(User user)
        {
            // Validate username uniqueness
            var exists = unitOfWork.Users.Find(u => u.Username == user.Username).Any();
            if (exists)
            {
                return new GenericResponse(false,
                    localizationService.GetLocalizedString("EntityAlreadyExists"));
            }

            // Persist
            await unitOfWork.Users.Add(user);
            return new GenericResponse(true, user);
        }

        public async Task<GenericResponse> CreateManagedUser(CreateManagedUserRequest request)
        {
            var errors = new List<string>();

            if (request.Password != request.RepeatPassword)
            {
                errors.Add(localizationService.GetLocalizedString("UserPasswordsDoNotMatch"));
            }

            var emailValidator = new EmailAddressAttribute();
            if (!emailValidator.IsValid(request.Email))
            {
                errors.Add(localizationService.GetLocalizedString("Validation.InvalidEmail"));
            }

            var existingUser = unitOfWork.Users.Find(u => u.Username == request.Username).FirstOrDefault();
            if (existingUser is not null)
            {
                errors.Add(localizationService.GetLocalizedString("UserNotAvailable", request.Username));
            }

            var role = await unitOfWork.Roles.Get(request.RoleId);
            if (role is null || role.Disabled)
            {
                errors.Add(localizationService.GetLocalizedString("RoleNotFound", request.RoleId));
            }

            var language = await languageCatalog.GetByCodeAsync(request.PreferredLanguage);
            if (language is null)
            {
                errors.Add(localizationService.GetLocalizedString("LanguageNotFound", request.PreferredLanguage));
            }

            if (errors.Count > 0)
            {
                return new GenericResponse(false, errors);
            }

            var user = new User
            {
                Username = request.Username,
                Password = BCrypt.Net.BCrypt.EnhancedHashPassword(request.Password),
                FirstName = request.FirstName,
                LastName = request.LastName,
                Email = request.Email,
                PreferredLanguage = request.PreferredLanguage,
                RoleId = request.RoleId,
                ProfileId = request.ProfileId,
                Disabled = false
            };

            await unitOfWork.Users.Add(user);
            return new GenericResponse(true, user);
        }

        public async Task<GenericResponse> UpdateUser(User user)
        {
            // Check if user exists
            var existing = await unitOfWork.Users.Get(user.Id);
            if (existing == null)
            {
                return new GenericResponse(false,
                    localizationService.GetLocalizedString("EntityNotFound", user.Id));
            }

            // Update
            await unitOfWork.Users.Update(user);
            return new GenericResponse(true, user);
        }
    }
}

File: frontend/src/services/user.service.ts
import { AxiosInstance } from "axios";
import apiClient, { logException } from "@/api/api.client";
import { User } from "@/types";

export class UserService {
  public apiClient: AxiosInstance;
  private resource: string;
  constructor() {
    this.apiClient = apiClient;
    this.resource = "/User";
  }

  public async GetAll(): Promise<User[] | undefined> {
    try {
      let response = await this.apiClient.get(this.resource);
      if (response.status === 200) {
        const models = response.data as Array<User>;
        return models;
      }
    } catch (error) {
      logException(error);
    }
  }

  public async GetById(id: string): Promise<User | undefined> {
    try {
      let response = await this.apiClient.get(`${this.resource}/${id}`);
      if (response.status === 200) {
        return response.data as User;
      }
    } catch (error) {
      logException(error);
    }
  }

  public async Update(user: User): Promise<boolean> {
    try {
      let response = await this.apiClient.put(
        `${this.resource}/${user.id}`,
        user
      );
      if (response.status === 200) {
        return true;
      }
    } catch (error) {
      logException(error);
    }
    return false;
  }

  public async CreateManaged(
    request: CreateManagedUserRequest,
  ): Promise<User | undefined> {
    try {
      const response = await this.apiClient.post(
        `${this.resource}/managed`,
        request,
      );
      if (response.status === 201 || response.status === 200) {
        return response.data as User;
      }
    } catch (error) {
      logException(error);
    }
  }
}

export interface CreateManagedUserRequest {
  username: string;
  password: string;
  repeatPassword: string;
  firstName: string;
  lastName: string;
  email: string;
  preferredLanguage: string;
  roleId: string;
  profileId?: string | null;
}
